### PR TITLE
refactor: remove bigger macros in favor of direct types

### PR DESCRIPTION
**Pull request by dignifiedquire**

### CODE REVIEW

File: src/composed/signed_key/key_parser.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+use crate::{SignedKeyDetails, SignedPublicSubKey, SignedSecretSubKey};
+
+#[allow(clippy::complexity)]
+pub fn next<I, IKT>(
+    packets: &mut std::iter::Peekable<I>,
+    key_tag: Tag,
+    parse_secrect_subkeys: bool,
+) -> Option<
+    Result<(
+        IKT,
```
Comment (hko-s) [inline src/composed/signed_key/key_parser.rs:13]: I'm unsure how to think about this generic.

I think the `IKT` should definitely never be a subkey type. Shouldn't it always be a `packet::PublicKey`?

Comment (hko-s) [reply src/composed/signed_key/key_parser.rs:13]: Ah. Or a `packet::SecretKey`?

Still, it seems to me the type system isn't currently guaranteeing that we won't get a subkey packet in this location? ... which would be a bit sad.

File: src/packet/key/public.rs [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+        Ok(Self(inner))
+    }
+
+    /// Parses a `PublicKeyKey` packet from the given slice.
+    pub fn from_slice(packet_version: Version, input: &[u8]) -> Result<Self> {
+        let inner = PubKeyInner::from_slice(packet_version, input)?;
+        Ok(Self(inner))
+    }
+
+    pub fn sign<F>(&self, key: &impl SecretKeyTrait, key_pw: F) -> Result<Signature>
```
Comment (hko-s) [inline src/packet/key/public.rs:45]: I wonder if this function (and its subkey sibling, below) should both be named `bind`, to be slightly more self-explanatory.

File: src/composed/signed_key/key_parser.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/composed/signed_key/key_parser.rs:13]: I am confused, it is quite strictly reduced type wise in the where clause, what are you worried about?

Comment (hko-s) [reply src/composed/signed_key/key_parser.rs:13]: I was wondering if it'd be possible and useful to disallow subkey types in this place. But I have no strong feelings about this.

**Comment by hko-s** [comment]
I haven't tried to verify that the transformations from macro to plain code are 1:1 the same, but assume they are.

That said, this PR seems like a great improvement to me - and using it as a dependency worked well, where i tried.

File: src/packet/key/public.rs (hunk 2 above)
Comment (hko-s) [reply src/packet/key/public.rs:45]: Moving this point to https://github.com/rpgp/rpgp/issues/367, it is separate from this PR

**Pull request merged**